x509_vfy.c: Make chain building more intuitive and complete

DDvO opened the pull request:
Add option (on by default) to stop chain building as early as possible, i.e, as soon as
* the current cert is in the trust store
* and its trust is not rejected
* and it is self-signed or partial chains are accepted.
 
Also improve `ossl_cmp_build_cert_chain()` using the new feature
and make it public under the name `OSSL_build_cert_chain()`.

Moreover, generalize the internal `check_trust()` to work not with leaf but with current cert.                                                         
This allows getting rid of strange exceptional cases in `find_issuer()`
(and maybe also in other parts of the overall chain building algorithm).
This fixes a regression introduced in and reported for version 1.1.1 in #13739.

Further related commits do the following:
* Add checks in case `X509_NAME_hash()` fails due to unsupported SHA1.
* Make `PEM_X509_INFO_read_bio_ex()` conservative on the error queue
* TEST: move cert, key, and CSR loading aux functions to new `testutil/load.c`

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

richsalz commented:
I am worried about changing behavior.  Please make the new behavior *off* by default.

DDvO commented:
> I am worried about changing behavior.

@richsalz can you make more concrete what you fear could go wrong here?
The change in question just prefers short chains during chain building, which improves its completeness.
The correctness of chain building is anyway checked later during chain verification, right?
Moreover, one should hope that the multitude of existing tests on chain building and verification covers all critical cases, and none of them fails after this change.
So I cannot see where a real danger should come in here.

DDvO commented:
The provided change happens to fix an issue that recently was reported on the openssl-users mailing list:

Someone wants to use a directly trusted self-signed SSL server cert 
and for some reason puts this cert in the list of trusted certs _after_ another version of this cert with same subject (and SAN extensions) but with a different public key.
So far, the chain building chose the other cert (because it matches first) and then failed to verify the server cert (also because no backtracking is done).
Instead, after the change provided here the chain building immediately takes the right server cert (found in second position) in the trusted list and verification succeeds.

>  Please make the new behavior off by default.

As the example just given shows, this would be counter-productive.

richsalz commented:
With this change, what is the need for the ```PARTIAL_CHAINS``` flag?

richsalz commented:
The fact that someone reported an issue does not outweigh the current users who are depending on the present semantics.

kroeckx commented:
Have you seen the discussion in #7871?

richsalz commented:
@kroeckx yes I have read the discussion in #7871. (My comment there raised the same issue as I did here: this is a behavior change happening very late in the 3.0 release process and I am opposed.) I wryly note @vdukhovni's comment, https://github.com/openssl/openssl/issues/7871#issuecomment-751446865, that @ddvo gave a thumbs-up to. :)

@ddvo, my question "what does this to the requiring PARTIAL_CHAINS" was directed to you.

This is a behavior change. There are no tests added that show the behavior change before and after. And when I pushed on that, the response was "one should hope that the multitude of existing tests on chain building and verification covers all critical cases, and none of them fails after this change."  HOPE IS NOT GOOD ENOUGH.  Here is a use-case that is not covered, for example: all sorts of certs are thrown into a trust store, including "trusted roots" and intermediaries. A few users, that want a particular path through a new intermediary, turn on PARTIAL_CHAINS. Consider when a CA adds a new digest mechanism (or crypto algorithm) that isn't supported on some devices.

Chain building and trust validation is hard, full of dark and mysterious corner cases, and changing it should be done upfront *EARLY* in the release plan, to socialize it (as Viktor said).  It should not be done *LATE* while we are trying to count the days to get the BETA out.

kroeckx commented:
@richsalz: My comment was directed at @DDvO, since I did not notice him participating in that issue, unlike you. I have to agree with most of what you say.

beldmit commented:
@openssl/otc, I think this PR deserves a special OTC discussion.

DDvO commented:
I became aware of the discussion in #7871 just a couple of hours back, and meanwhile I've added to two major contributions to it.
In the light of that discussion I've started refining this PR w.r.t the treatment of partial chains...

DDvO commented:
> With this change, what is the need for the `PARTIAL_CHAINS` flag?

This PR does not change the semantics of `X509_V_FLAG_PARTIAL_CHAIN`,
which is a different issue, still under discussion in #7871.
The improvements done here just make sure that more valid chains are found
(assuming the same set of trust anchors as before).
In other words, it does not change trust policies but increases chain building completeness.

DDvO commented:
> The fact that someone reported an issue does not outweigh the current users who are depending on the present semantics.

Sure, but I still doubt that someone is depending on the _**incompleteness**_ of the hitherto implementation.
A counterexample would convince me of the contrary. @richsalz, the challenge to provide one is still open :wink:

DDvO commented:
> There are no tests added that show the behavior change before and after.

I've meanwhile added the example discussed recently on the openssl-users list that I mentioned above:
https://www.mail-archive.com/openssl-users@openssl.org/msg88961.html

It demonstrates the effectiveness of the enhancements done here for an extremely simple case.
Conversely, the hitherto implementation had been failing even in such extremely simple cases.

richsalz commented:
It will take me a couple of days of elapsed time to run some experiments.

richsalz commented:
> Sure, but I still doubt that someone is depending on the incompleteness of the hitherto implementation.

You are wrong.  Countless people are no doubt depending, in the fact of lack of documentation otherwise, and the various implementation quirks of OpenSSL.  The project has, rightly so, been very cautious to change existing behavior even when it's a complete pain in the neck.  See, for example, the recent discussions about initializing keys, private keys without public, etc.

kroeckx commented:
The changes to x509 code in the last 2 stable releases each time broke several applications.

DDvO commented:
> > Sure, but I still doubt that someone is depending on the incompleteness of the hitherto implementation.
> 
> You are wrong. Countless people are no doubt depending, in the fact of lack of documentation otherwise, and the various implementation quirks of OpenSSL. 

@richsalz, your combination of harsh rejection and weak/superficial argumentation is not acceptable, at least to me.

Again, why on earth should anyone rely on existing incomplete chain building? You understand what this means?
As long as you are unable to provide a real concrete point here it makes little sense to continue discussing with you.

> The project has, rightly so, been very cautious to change existing behavior even when it's a complete pain in the neck.

I see the general value of this point, but cannot see any implication against making chain building _more complete_.
Of course, one needs be careful to avoid introducing new bugs, which effectively would cause _less_ coverage/completeness regarding other aspects of chain building. But this would be a bad argument, because it could be used against doing _any_ improvements just because they might introduce new bugs.

> See, for example, the recent discussions about initializing keys, private keys without public, etc.

This is a different topic, and analogy arguments are very weak if not wrong, in particular given the complexity of the topics.

DDvO commented:
> The changes to x509 code in the last 2 stable releases each time broke several applications.

Which changes do you mean exactly?
Weren't those instances of the more general point I've just given above,
namely that of course every improvement has the potential to introduce new bugs?
Of course, they should be avoided, and if they happen should be detected and must be fixed ASAP.

I am not aware of any change regarding x509 that fundamentally broke something or even stronger: 
that (potentially after some minor fixing) did not result in a pure improvement of the overall situation.

beldmit commented:
At least there was a bug caused a lack of diagnostics when the root certificate was expired

DDvO commented:
> At least there was a bug caused a lack of diagnostics when the root certificate was expired

Indeed. This is a very special and interesting example because there, 
while fixing some other aspect, I - accidentally - removed a feature that
* is not required by RFCs and handled differently by other implementations, so is at least debatable
* is agreeably nevertheless good practice
* was lacking a (negative) test case
* was so far not (clearly) documented, which I've taken care to fix

This examples falls into the category I just mentioned above: 
It was found externally rather soon and was fixed immediately after being reported.
So in the end (after that fixup) the overall situation was improved.

beldmit commented:
@DDvO , I have no attention to blame you!

I'm seriously afraid of changes of the chain building behavior because it is messy by set of the standards, messy by implementation, and hardly can be debugged by test - only in "black box" mode. Those reasons make me (and I'm sure @richsalz  has the similar motivation) nervous when any changes, even the improving ones, are introduced.

DDvO commented:
Yeah, unfortunately the whole cert chain building and verification situation is indeed pretty messy.
I've already started writing up a meta issue about it yesterday, and will be able to post it later today.
I suggest placing all the more general discussion there.

kroeckx commented:
> Which changes do you mean exactly?

The details aren't really important. I think every stable release has
introduced regressions. But to give an example, the last release
changed is so that the expiration time now gets checked if you
ignore the error that the the CA is not found. This results in an
additional callback that confuses some applications. It also means
that applications that used it to ignore the expiration can stop
working because now they get a error they didn't know they had to
ignore.

> Weren't those instances of the more general point I've just given above,
> namely that of course every improvement has the potential to introduce new bugs?

It's not just introducing bugs, but changing behaviour that other
applications depend on. Every bug is someone's feature?

Some times it just a test suite that starts to fail, because it
checked that we did the old behaviour. Other times applications
stop working. Some of the changes actually point to security
issues in the applications.

All I'm saying is that we need to be careful what we change in
stable releases.

t8m commented:
> It's not just introducing bugs, but changing behaviour that other applications depend on. Every bug is someone's feature? Some times it just a test suite that starts to fail, because it checked that we did the old behaviour. Other times applications stop working. Some of the changes actually point to security issues in the applications. All I'm saying is that we need to be careful what we change in stable releases.

I am also partially responsible for those breaking changes as I approved them and I regret that. At least for such critical and complex parts of the code base as the X509 chain building and verification is we should really avoid touching the released branch unless there is a truly serious issue found. All the changes here should bake for sufficiently long time in the master branch and only after they are well tested (at the earliest after the 3.0 beta is released and tested) we should consider backporting them to 1.1.1 if they fix real bugs.

DDvO commented:
I agree that one needs to be particular careful with making changes to stable versions (which BTW is not the intention of this PR). 

A further complication with this, which I recently felt more than once, is that (naturally) external parties tend to report bugs on released versions such as 1.1.1., while they typically need to be fixed for the master (currently 3.0.0-pre) as well and then need to be back-ported. 
In particular in the latter step, recently there were a couple of cases where I was tempted to do more than would have been barely needed, which does come with the risk of introducing regressions.

richsalz commented:
@ddvo I am sorry you feel I was being too harsh to continue discussing this.  I will respect your wishes.

DDvO commented:
> @DDvO I am sorry you feel I was being too harsh to continue discussing this. I will respect your wishes.

Thanks @richsalz. I did not mean to stop discussion in general, but just exchanging superficial statements that do not take into account the actual contents of the PR. If you can point out any concrete problem with it, I'd be happy to learn about this.

richsalz commented:
```
        Sure, but I still doubt that someone is depending on the incompleteness of the hitherto implementation.

    You are wrong. Countless people are no doubt depending, in the fact of lack of documentation otherwise, and the various implementation quirks of OpenSSL.

@richsalz, your combination of harsh rejection and weak/superficial argumentation is not acceptable, at least to me.
```

I don't understand how my claim is any more weak or superficial than your yours. In fact, my claim is based on decades of experience with OpenSSL, including many years as a member of the project management team.  It has been stated multiple times, by team members, in various forums and discussions, "we don't know how openssl is being used."  Other people have posted on this PR about cases where changes broke things.

```
    See, for example, the recent discussions about initializing keys, private keys without public, etc.

This is a different topic, and analogy arguments are very weak if not wrong, in particular given the complexity of the topics.
```

No, it's the same thing.  All of these are instances of people relying on corner cases and poorly documented behavior and saying "you'll break my code" when the project tries to "fix" or "clean things up."  Don't take it so literally; changing ECDSA key behavior to require both public and private is *the same thing*, to the user, as changing how trust stores works.

There was also no response to the specific questions in the third paragraph of my comment at https://github.com/openssl/openssl/pull/13748#issuecomment-753478290 .

I just noticed that (either three or seven days ago, not clear when) that you now believe this is post-3.0.  That's what I wanted, so thanks.

richsalz commented:
> It will take me a couple of days of elapsed time to run some experiments.

As this is no longer planned for 3.0, I'm not planning on doing this in the near term.  (See https://github.com/openssl/openssl/pull/13748#issuecomment-754062497 and the the entries just before it for context.)

DDvO commented:
>>>> Sure, but I still doubt that someone is depending on the incompleteness of the hitherto implementation.
>>>  
>>>  You are wrong. Countless people are no doubt depending, in the fact of lack of documentation otherwise, and the various implementation quirks of OpenSSL.
>> 
>> @richsalz, your combination of harsh rejection and weak/superficial argumentation is not acceptable, at least to me.
> 
> I don't understand how my claim is any more weak or superficial than your yours. In fact, my claim is based on decades of experience with OpenSSL, including many years as a member of the project management team. It has been stated multiple times, by team members, in various forums and discussions, "we don't know how openssl is being used." Other people have posted on this PR about cases where changes broke things.

I consider your claim superficial frankly because you do not argue based on a real understanding what this PR does.
You'd need to invest considerable effort and time into this specific matter, which you have not done.

>>>  See, for example, the recent discussions about initializing keys, private keys without public, etc.
>> 
>> This is a different topic, and analogy arguments are very weak if not wrong, in particular given the complexity of the topics.
> 
> No, it's the same thing. All of these are instances of people relying on corner cases and poorly documented behavior and saying "you'll break my code" when the project tries to "fix" or "clean things up." Don't take it so literally; changing ECDSA key behavior to require both public and private is _the same thing_, to the user, as changing how trust stores works.

As mentioned, of course I agree that one must be very careful when changing things 
as complex, brittle, messy, and poorly documented as the chain building implementation.
But fear of breaking things must not imply that improvements cannot be done. 
I believe your fears would be less if you had taken the time to understand what the changes proposed here actually do.

> There was also no response to the specific questions in the third paragraph of my comment at [#13748 (comment)](https://github.com/openssl/openssl/pull/13748#issuecomment-753478290) .
 
No, https://github.com/openssl/openssl/pull/13748#issuecomment-754061164 was a response to it, and to that *your* response is missing.
That paragraph does not contain a question, but anyway here is some more explicit responses on it:

>>> There are no tests added that show the behavior change before and after. 

This has changed on Jan 4th when I added a test case that would not have succeeded before:
`"accept ee-cert2 although matching also first cert in ee-certs"`.

>>> And when I pushed on that, 

Here you were twisting the timeline. My general comment on existing tests were *before* your above your request for a specific tests confirming that this PR brings improvements.

>>> Here is a use-case that is not covered, for example: [...]

Correct that I did not yet react on this example, which is rather fuzzy - would be good if you formalized it as a test case.
I will try to understand it and comment on it accordingly in my next comment, below.

> I just noticed that (either three or seven days ago, not clear when) that you now believe this is post-3.0. That's what I wanted, so thanks.

All right.

DDvO commented:
@richsalz wrote:

> Here is a use-case that is not covered, for example: all sorts of certs are thrown into a trust store, including "trusted roots" and intermediaries. A few users, that want a particular path through a new intermediary, turn on PARTIAL_CHAINS. Consider when a CA adds a new digest mechanism (or crypto algorithm) that isn't supported on some devices.

So in this scenario there is a path through an intermediate CA cert using an older signature (digest/signing) algorithm
and an alternative path through another cert for the same intermediate CA with a new signature algorithm, 
and there is a shared trust store that contains both these certs as non-trusted, as well as trusted roots for both of them, right?

Before turning on PARTIAL_CHAINS for that trust store, 
if a cert verification implementation supports only the older algorithm, it will succeed with taking the older path;
else it will succeed with choosing either the older or the new path. 

After turning on PARTIAL_CHAINS for that trust store, 
if a cert verification implementation supports only the older algorithm, it will succeed with taking the older path chopped at the  cert for that intermediate CA;
else it will succeed with choosing either the older or the new path chopped at the respective cert for that intermediate CA.

In both cases, the changes done by this PR won't change this behavior.
At least, this would not be intended, since it just aims at preferring short paths that, as before, need to be verifiable using the _available_ algorithms. The latter subtlety is implemented as part of `check_sig_alg_match()` in `v3_purp.c`:
```
    if (OBJ_find_sigid_algs(OBJ_obj2nid(subject->cert_info.signature.algorithm),
                            NULL, &pkey_nid) == 0)
        return X509_V_ERR_UNSUPPORTED_SIGNATURE_ALGORITHM;
```
and has so far been missing from my recent improvement to `openssl-verification-options.pod`:
>> A candidate issuer certificate matches a subject certificate
>> if all of the following conditions hold:
>> * Its subject name equals the issuer name of the subject certificate.
>> * If the subject certificate has an authority key identifier extension,
>> each of its sub-fields equals the corresponding subject key identifier, 
>> serial number, and issuer field of the candidate issuer certificate,
>> as far as the respective fields are present in both certificates.
>> * The certificate signature algorithm used to sign the subject certificate
>> equals the public key algorithm of the candidate issuer certificate.

I've just extended the last item to:
>> * The certificate signature algorithm used to sign the subject certificate
>> is supported and
>> equals the public key algorithm of the candidate issuer certificate.

vdukhovni commented on crypto/x509/x509_vfy.c:
```diff
@@ -2317,6 +2286,10 @@ int X509_STORE_CTX_init(X509_STORE_CTX *ctx, X509_STORE *store, X509 *x509,
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_X509, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
```
This appears unrelated to the PR.

DDvO replied:
Right, it's a "drive-by improvement".

DDvO replied:
I've meanwhile moved this unrelated change to a different PR.

vdukhovni commented on crypto/x509/x509_vfy.c:
```diff
@@ -3112,6 +3085,11 @@ static int build_chain(X509_STORE_CTX *ctx)
             }
             x = sk_X509_value(ctx->chain, i-1);
 
+            /* Stop with success if current cert equals a trust anchor */
+            if (lookup_cert_match(ctx, x) != NULL
+                    && (trust = check_trust(ctx, num)) == X509_TRUST_TRUSTED)
+                break;
+
```
I don't believe this new code is needed, or appropriate.  The exact match lookups are only appropriate for EE certs and already handled in check_trust() as fallback when partial chains are enabled.

DDvO replied:
See my above comment to https://github.com/openssl/openssl/pull/13748#discussion_r555289273.

vdukhovni replied:
Sorry, this code is not correct, because presumptively we have a chain that's not yet complete, and the very next line looks for an issuer of the most recently added untrusted certificate, that code also checks whether the newly added issuer is trusted or not.  Other than perhaps with an EE cert at depth 0, this check looks redundant, because the trust was checked when the cert was added to the chain and does not need to be checked again.

DDvO replied:
Sorry, it **is** correct and even required because as mentioned there are situations where it does not even make sense to try looking for an issuer cert.

Meanwhile I've simplified the new code at this point to:
```
            if ((trust = check_trust(ctx, num)) == X509_TRUST_TRUSTED)
                break;
```
It is required because otherwise the following new test case `"accept ee-cert although deceptively matching also first cert in ee-certs"` would fail like this:
```
apps/openssl verify -auth_level 1 -purpose any -no_check_time -trusted test/certs/ee-noKID-certs.pem test/certs/ee-noKID-cert2.pem
C = US, ST = CA, L = Palo Alto, O = VMware, CN = nsxmanager.pks.vmware.local
error 18 at 0 depth lookup: self-signed certificate
error test/certs/ee-noKID-cert2.pem: verification failed
```
Moreover, the test case `"accept last-resort direct leaf match Ed25519-signed self-issued cert"`:
`verify -partial_chain -trusted test/certs/ee-ed25519.pem test/certs/ee-ed25519.pem`
would else produce the spurious intermittent error `X509_V_ERR_SIGNATURE_ALGORITHM_MISMATCH` just because the chain building needlessly tries to find an issuer cert, which is found in the trust store but cannot be used for signing.

vdukhovni reviewed (changes requested):
Some more work required

DDvO commented on crypto/x509/x509_trs.c:
```diff
@@ -242,10 +242,14 @@ static int trust_compat(X509_TRUST *trust, X509 *x, int flags)
+    if ((x->ex_flags & EXFLAG_SS) != 0) {
+        if ((flags & X509_TRUST_NO_SS_COMPAT) == 0)
+            return X509_TRUST_TRUSTED;
+    } else {
```
It depends on the exact purpose of X509_TRUST_NO_SS_COMPAT, which is not documented.
If it is meant to strictly reject trust in self-signed certs (which is what I assumed here) the code is correct, but if the idea is that X509_V_FLAG_PARTIAL_CHAIN overrides every settings regarding trust in self-signed certs then of course the order of checking X509_TRUST_NO_SS_COMPAT and X509_V_FLAG_PARTIAL_CHAIN needs to be flipped.
So far I had understood that the default trust compatibility for self-signed certs and for "partial chains" are meant to complement each other, but if "partial chains" should take preference we can easily accommodate this here as well.

vdukhovni replied:
The NO_SS_COMPAT flag drops implicit trust-anchor semantics for self-signed certs in the trust store.  But, if with partial chain, we say that any cert in the trust store is a trust-anchor, even without explicit trust EKUs, then I really don't see why we should discriminate against the self-signed ones, and accept only intermediates.  That feels rather surprising to me.

DDvO replied:
I see, so `NO_SS_COMPAT` is not meant as strict as the name suggests,
and I'm going to adapt the handling to give X509_V_FLAG_PARTIAL_CHAIN preference.

DDvO replied:
Done.

DDvO commented on crypto/x509/x509_vfy.c:
```diff
@@ -146,8 +148,6 @@ static X509 *lookup_cert_match(X509_STORE_CTX *ctx, X509 *x)
         xtmp = NULL;
     }
-    if (xtmp != NULL && !X509_up_ref(xtmp))
-        xtmp = NULL;
```
I've changed this is because this PR uses `lookup_cert_match()` also directly in `build_chain() ` in a way where there is no need to keep a reference to the result.
Also the existing use in `check_trust()` becomes simpler if the `up_ref()` is done there in case really needed, rather than freeing the reference again in case it turns out it is not actually needed.
So in essence this is an internal code cleanup.

vdukhovni replied:
I don't presently know why the latter change is needed. We should probably figure that out first.

DDvO replied:
Yes, although, as mentioned, the proposed move of `up_ref()` to where really needed is also more generally a code simplification.

vdukhovni replied:
This change is wrong. It is not safe to remove the up_ref, while still returning the certificate, which may be freed before the function returns by the line below, which is needed to free the rest of the list.

DDvO replied:
Well, if `ctx->lookup_certs` is `lookup_certs_sk()` or `X509_STORE_CTX_get1_certs()`, then the reference counter of the returned certs is certainly >1, so the cert returned, `xtmp`, cannot actually be freed prematurely by `sk_X509_pop_free(certs, X509_free)`.
But still good point, there is the risk that a function set via `X509_STORE_set_lookup_certs()` does not guarantee this.
So reverted this change.

DDvO commented on crypto/x509/x509_vfy.c:
```diff
@@ -325,16 +325,12 @@ static X509 *find_issuer(X509_STORE_CTX *ctx, STACK_OF(X509) *sk, X509 *x)
-            if (x509_check_cert_time(ctx, rv, -1))
-                break;
+        if (ctx->check_issued(ctx, x, issuer)
+                && !sk_X509_contains(ctx->chain, issuer)) {
```
The case you are worried about is now much more elegantly handled by the part calling `lookup_cert_match()` directly in `build_chain()`. This is what I meant in the beginning of the description of this PR, and also wrote in #13780, by:

... fixes a design flaw of `build_chain()`:
it makes little sense to try finding an issuer of the current end of the chain if this end can be found immediately in the trust store (and is actually trusted).

vdukhovni replied:
What "design flaw" has chosen semantics, matching previously expected behaviour.  I don't yet understand the impact of the change, so I'll have to stare at it harder, but for now I remain sceptical that this is needed. Also, what is the performance cost?  Are we now doing more lookups while building the chain?

DDvO replied:
The proposed change to check for a direct trust store match as soon as possible during the executing of `build_chain()` not only avoids earlier hick-ups (there may be needless errors when trying to find an issuer cert that is not really needed) but also leads to less lookups and more efficiency.

DDvO commented on crypto/x509/x509_vfy.c:
```diff
@@ -824,7 +820,8 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
         x = sk_X509_value(ctx->chain, i);
-        trust = X509_check_trust(x, ctx->param->trust, 0);
+        trust = X509_check_trust(x, ctx->param->trust,
+                                 ctx->param->flags & X509_V_FLAG_PARTIAL_CHAIN);
```
I see. This won't be hard to correct.

DDvO replied:
Done.
BTW, it looks like one could simply reuse `X509_TRUST_OK_ANY_EKU` here, right?
Yet for now for clarity I've introduced `X509_TRUST_DO_PC_COMPAT`,

kaduk replied:
My undestanding was that `X509_TRUST_OK_ANY_EKU` is to indicate how to handle the special "anyExtendedKeyUsage" KeyPurposeId, which has the very curious behavior of:
```
   Applications that require the presence of a
   particular purpose MAY reject certificates that include the
   anyExtendedKeyUsage OID but not the particular OID expected for the
   application.
```

DDvO replied:
Interesting explanation, but I just had a closer look at the implementation and it turns out that the semantics of `X509_TRUST_OK_ANY_EKU` is very different:

First of all, the `anyExtendedKeyUsage` extended key usage OID is recognized by `x509v3_cache_extensions()`, setting `XKU_ANYEKU` in the cert structure, but apart from that `anyExtendedKeyUsage` entries are **ignored**!

Independently of that, if `X509_TRUST_OK_ANY_EKU` is set in the verification flags, it leads to `X509_TRUST_REJECTED` if the explicit rejection list in the auxiliary info f a (trusted) certificate contains `anyExtendedKeyUsage` (which makes the name `X509_TRUST_OK_ANY_EKU` rather misleading),
and it leads to `X509_TRUST_TRUSTED` if the explicit trust list  in the auxiliary info of a (trusted) certificate contains `anyExtendedKeyUsage`.

So in a nutshell, `X509_TRUST_OK_ANY_EKU` governs whether `anyExtendedKeyUsage` in any auxiliary trust info of a trust store element is recognized as trust wildcard trust or distrust wildcard, respectively.

DDvO replied:
Isn't it a shame that there is so much unclarity even among experts what various verification options do?
The `XKU_ANYEKU` is mentioned in the documentation but the only thing stated is only said that it corresponds to `anyExtendedKeyUsage` - and even this is misleading because this (pseudo-)EKU entry gets ignored.
And `X509_TRUST_OK_ANY_EKU` is not even mentioned in the documentation.

kaduk replied:
Thank you for digging into this more.

> So in a nutshell, X509_TRUST_OK_ANY_EKU governs whether anyExtendedKeyUsage in any auxiliary trust info of a trust store element is recognized as trust wildcard trust or distrust wildcard, respectively.

I am not sure whether I am matching things up as you intended with the "respectively".  In my reading, `X509_TRUST_OK_ANY_EKU` must be set in the flags in order for anyExtendedKeyUsage to be processed "at all" (by `obj_trust()`).  The behavior when anyExtendedKeyUsage is present (and the flag is set) depends on whether the OID is present in the 'reject' or 'trust' attributes.

> Isn't it a shame that there is so much unclarity even among experts what various verification options do?

Yes.  This lack of clarity, and the under-documented nature of many things, leads many of us to be very reluctant to introduce behavior changes...

DDvO replied:
Yes, in other words: `obj_trust()` ignores any `anyExtendedKeyUsage` entry in the trust or reject settings unless `X509_TRUST_OK_ANY_EKU` is set.
If `X509_TRUST_OK_ANY_EKU` is set, `anyExtendedKeyUsage` in the reject list causes all purposes to be rejected (which is not properly reflected in the name  `X509_TRUST_OK_ANY_EKU`), and
if `X509_TRUST_OK_ANY_EKU` is set, `anyExtendedKeyUsage` in the trust list causes all purposes to be accepted.
This should be clearly documented.
So far, there is no documentation except for the following comment in `x509.h.in`:
```
/* Accept "anyEKU" as a wildcard trust OID */
```
As a quick initial fix I've just extended this comment as:
```
/* Accept "anyEKU" as a wildcard rejection OID and as a wildcard trust OID */
```
within #13735.
```

DDvO commented on crypto/x509/x509_vfy.c:
```diff
@@ -833,41 +830,26 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
-        if (trust == X509_TRUST_REJECTED) {
-            X509_free(mx);
+        trust = X509_check_trust(mx, ctx->param->trust,
+                                 ctx->param->flags & X509_V_FLAG_PARTIAL_CHAIN);
```
Yep

DDvO replied:
Done

DDvO commented on crypto/x509/x509_vfy.c:
```diff
@@ -833,41 +830,26 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
-            return X509_TRUST_UNTRUSTED;
+    x = sk_X509_value(ctx->chain, num_untrusted - 1);
+    if (num_untrusted == num
+            && (mx = lookup_cert_match(ctx, x)) != NULL) {
```
This needs to be interpreted in the context of the above mentioned core change of this PR, namely adding to `build_chain()`:
```
/* Stop with success if current cert equals a trust anchor */
            if (lookup_cert_match(ctx, x) != NULL
                    && (trust = check_trust(ctx, num)) == X509_TRUST_TRUSTED)
                break; 
 ```
 If there a direct match of the current cert (i.e., the current end of the chain being built) with a trusted cert (no matter if it's a not self-signed EE cert or any other sort of cert) then we can succeed immediately.
Maybe the generality of having potentially some intermediate certs before is not needed anymore?

DDvO replied:
I've had a look at this again.
The 'last resort' case (for an EE cert that is in the trust store) that you mention above:
> directly match an EE cert that is not self-signed and comes along with an untrusted issuer chain.

is covered by the new, generalized code because due to the new call `(trust = check_trust(ctx, num)) == X509_TRUST_TRUSTED` in `build_chain()`, `check_trust()` gets called for each trust store element added to the chain, and  when called for the first time `num_untrusted` will be `1`, so the EE (leaf) cert will be handled as desired.
This is witnessed by the test cases `"accept last-resort direct leaf match"` which still pass.

Note that in the PKIX case the check commented as `Last chance to make a trusted chain,
either bare DANE-TA public-key signers or direct leaf PKIX trust.`
after the loop is now not needed anymore, so I've removed it.

Moreover, the code section in build_chain()`:
```
                } else if (num == ctx->num_untrusted) {
                    /*
                     * We have a self-signed certificate that has the same
                     * subject name (and perhaps keyid and/or serial number) as
                     * a trust anchor.  We must have an exact match to avoid
                     * possible impersonation via key substitution etc.
                     */
                    if (X509_cmp(curr, issuer) != 0) {
                        /* Self-signed untrusted mimic. */
                        X509_free(issuer);
                        ok = 0;
                    } else { /* curr "==" issuer */
                        X509_free(curr);
                        ctx->num_untrusted = --num;
                        (void)sk_X509_set(ctx->chain, num, issuer);
                        curr = issuer;
                        /* no need to update self_signed */
                    }
                }

```
was used only in two degenerate cases:
 * for direct trust of EE cert
 * if X509_V_FLAG_TRUSTED_FIRST unset and trust anchor is also in untrusted

I've disabled it as well because it is covered by the generalized `check_trust()` now called before,
which checks direct trust for current cert (at index num_trusted - 1).

All this did not lead to a change in documented behavior.

DDvO commented on crypto/x509/x509_vfy.c:
```diff
@@ -1743,37 +1725,24 @@ static int internal_verify(X509_STORE_CTX *ctx)
-            goto check_cert_time;
-        }
-        if (n <= 0) {
-            CHECK_CB(1, ctx, xi, 0, X509_V_ERR_UNABLE_TO_VERIFY_LEAF_SIGNATURE);
```
If the chain building yields a chain where the last cert in the chain is not self-issued (regardless whether it's an EE or CA cert), this can only happen if that cert is trusted and X509_V_FLAG_PARTIAL_CHAIN is active, right?
Then there is no point in trying to verify its signature, as I wrote in the new comment.

vdukhovni replied:
Well, chain building can also fail to construct a trusted chain.  This callback was raising an error that the application verify callback could optionally override. That error does not seem to be raised anymore, so we need to be sure that's OK.

DDvO replied:
In case the chain building cannot find a path that ends in a trust anchor it must not declare success.
So the chain verification (and thus the trigger for that callback) should not be reachable at all in this case!

DDvO replied:
BTW, to be on the safe side I advise to make `internal_verify()` (double-)check that the end of the chain presented to it is trusted. This should be done not only in the special case of a not self-issued cert but generally. This improvement is proposed in #13770.

vdukhovni replied:
That might break DANE with bare public key, though I probably have tests for that.  Caution advised.  And the check is not entirely free, I am not sure it is worth it.

DDvO replied:
I have essentially zero knowledge of DANE, so would be great if you can check the DANE implications - maybe the approach needs to adapted to this end.

vdukhovni replied:
Yes, tests `#44` and `#45` in `test/danetest.in` check bare public key trust-anchor support.

DDvO replied:
And the tests still pass :)

DDvO commented:
(Only) the Windows versions of the tests keep failing like this:
```
Could not open file or uri for loading certs of trusted certificates from ..\..\..\..\test\recipes\25-test_verify_data\ee-certs.pem
38080000:error:16000069:STORE routines:ossl_store_get0_loader_int:unregistered scheme:..\crypto\store\store_register.c:240:scheme=file
38080000:error:80000002:system library:file_open:No such file or directory:..\providers\implementations\storemgmt\file_store.c:278:calling stat(..\..\..\..\test\recipes\25-test_verify_data\ee-certs.pem)
C:\Perl\bin\perl.exe ..\..\util\wrap.pl ..\..\apps\openssl.exe verify -auth_level 1 -purpose any -trusted ..\..\..\..\test\recipes\25-test_verify_data\ee-certs.pem ..\..\..\..\test\recipes\25-test_verify_data\ee-cert1.pem => 1
not ok 154 - accept ee-cert1 matching first cert in ee-certs
```
because `test\recipes\25-test_verify_data\ee-certs.pem` cannot be found although the file must be there.
On all other platforms `data_file()` works fine - does anyone have a hint why this goes wrong for Windows
and how to rectify this?

t8m commented on include/openssl/x509.h.in:
```diff
@@ -1025,6 +1025,9 @@ int EVP_PKEY_add1_attr_by_txt(EVP_PKEY *key,
                               const char *attrname, int type,
                               const unsigned char *bytes, int len);
 
+STACK_OF(X509) *OSSL_build_cert_chain(OSSL_LIB_CTX *libctx, const char *propq,
```
Do we really want this to be a public API?

DDvO replied:
So far it is not strictly needed but may be of more general interest to certain applications.
For instance, mbedTLS export such a function.
So far the chain built by OpenSSL can be extracted only in conjunction with chain verification,
which may be more than needed/wanted for some use cases.

DDvO replied:
I've meanwhile taken the addition of a generally useful chain building function to a separate PR: #14128.

t8m commented on doc/man3/X509_verify_cert.pod:
```diff
@@ -37,11 +48,18 @@ Applications must check for <= 0 return value on error.
-If a complete chain can be built and validated this function returns 1,
-otherwise it return zero, in exceptional circumstances it can also
-return a negative code.
+ossl_cmp_build_cert_chain() returns NULL on error,
```
Old name used here.

DDvO replied:
Oops, fixed.

vdukhovni commented:
One more thing to think about in terms of partial-chain is that with a partial chain one might fail to take into account name constraints that apply to an issuer CA higher up the chain, EKU restrictions on a trust-anchor further up the chain, or even (rarely used) evaluation of "policy" OIDs (for which are there optional tests one can run against the NIST test harness, that I've run a few times, but don't recall exactly how to make it go with any level of detail).

I still think that we'd be better off investigating support for an intermediate (untrusted) certificate store, and ideally explicit trust EKUs for all trust-anchors, which would make trust-intermediates explicit, rather than implicit via partial-chain. And also would facilitate more explicit definitions of what each trust-anchor is trusted for...

vdukhovni commented:
I haven't looked at this PR in a little while.  There are some good bits here IIRC and some things here and some more controversial. Perhaps you could stage just the parts that are cleaning up code and not changing policy into a smaller separate PR, and we can then think about just the policy changes separately, if you're still keen on those too?

DDvO commented:
Yeah I'll do that (as soon as I find the time for it).

DDvO commented:
> I haven't looked at this PR in a little while. There are some good bits here IIRC and some things here and some more controversial. Perhaps you could stage just the parts that are cleaning up code and not changing policy into a smaller separate PR, and we can then think about just the policy changes separately, if you're still keen on those too?

@vdukhovni, I've meanwhlle carved out all unrelated improvements,
fixed the minor issues you pointed out here,
updated/corrected/extended some comments, and
re-visited the topics discussed above (see there for details).

In a nutshell: 
I'm meanwhile more convinced than ever that the changes proposed here are correct and even needed for improving completeness and to avoid inefficient and even counterproductive extra checks (which can lead also to spurious errors).
Moreover, the code has become a little cleaner and thus better to read and maintain.

richsalz commented:
@ddvo, does your latest comment mean you want this in 3.0?

DDvO commented:
> @DDvO, does your latest comment mean you want this in 3.0?

Not really - I did not change the milestone, which still is **Post 3.0.0**.
I do not see the proposed improvements as urgent
but just wanted to update the contents of this PR
and give the possibility to advance the discussion on it.

vdukhovni commented on crypto/x509/x509_vfy.c:
```diff
@@ -3216,25 +3220,23 @@ static int build_chain(X509_STORE_CTX *ctx)
-    }
+    if (trust == X509_TRUST_UNTRUSTED && num <= depth
+        && DANETLS_HAS_DANE_TA(dane))
+        /*
+         * Last chance to make trust for bare DANE-TA public-key signers
+         */
```
```suggestion
        /* Last chance to make trust for bare DANE-TA public-key signers */
```

DDvO replied:
Done

DDvO commented:
Rebased to fix a merge conflict.
A few further adaptations and TODOs added.

DDvO commented on crypto/x509/x509_vfy.c:
```diff
@@ -3116,31 +3112,39 @@ static int build_chain(X509_STORE_CTX *ctx)
                         ctx->error = X509_V_ERR_OUT_OF_MEM;
                         break;
                     }
+#if 0 /* This part was used only in two degenerate cases:
```
This comment is only preliminary (because I think the whole `#if 0` ... `#endif` part can be dispensed with), but re-formatted it anyway.

DDvO commented on crypto/x509/x509_vfy.c:
```diff
@@ -796,57 +796,53 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
+     * possible impersonation via key substitution etc.
+     */
+    if (num_untrusted == num
+        && (mx = lookup_cert_match(ctx, x)) != NULL) {
```
As I wrote in the description of this PR:
> it makes little sense to try finding an issuer of the current end of the chain if this end can be found immediately in the trust store (and is actually trusted).

vdukhovni replied:
Other than the EE cert, we always know the provenance of each subsequent certificate as we're building the chain, because we added it, either by getting it from the trust store or from the untrusted store (often peer-provided chain). This is tracked by virtue of the "num_untrusted" variable.  Therefore, I still fail to see the point (which you're failing to explain precisely in a non hand waving way) of applying the EE-special trade-in for a trust-store equivalent at any other depth in the chain.

You just have to explain this much better.

DDvO replied:
Good point that for each cert in the chain being constructed, except for the target cert (which BTW does not need to be an EE cert) we already know whether its is from the trust store, and in fact we already call `check_trust()` on the current cert if it is from the trust store, and stop the loop if its is either trusted or rejected!
This was not clear to me because the whole chain building loop is so spread-out and complicated with detail (which partly could and better should be abstracted away).
So thanks to your comment I meanwhile see that for the certs after the target cert the new check is actually redundant!

vdukhovni replied:
> Good point that for each cert in the chain being constructed, except for the target cert (which BTW does not need to be an EE cert) we already know whether its is from the trust store, and in fact we already call `check_trust()` on the current cert if it is from the trust store, and stop the loop if its is either trusted or rejected!
> This was not clear to me because the whole chain building loop is so spread-out and complicated with detail (which partly could and better should be abstracted away).
> So thanks to your comment I meanwhile see that for the certs after the target cert the new check is actually redundant!

We're making progress!  If you limit the new check for just depth 0, with a suitable comment, my objections may be resolved.  You're then just handling EE direct match for partial chain early, skipping the chain construction. That's probably OK, if that's what you're after.  I can support that change, building a more complete chain seems redundant (the caller asked for partial chains, *and* has the EE cert in the store, could they possible indicate intent any more clearly???)

DDvO replied:
Done this, as soon as possible, as follows:
```
/*
 * Stop immediately if target cert is directly trusted/rejected.
 * This improves completeness and prevents inefficiency and spurious
 * errors, which can occur when it makes no sense to look for an issuer.
 */
if (num == 1 && (trust = check_trust(ctx, 1)) != X509_TRUST_UNTRUSTED)
    break;
```

DDvO commented on crypto/x509/x509_vfy.c:
```diff
@@ -796,57 +796,53 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
-        /* Replace leaf with trusted match */
-        (void)sk_X509_set(ctx->chain, 0, mx);
-        X509_free(x);
-        ctx->num_untrusted = 0;
```
Yes, this is now done within check_trust()` - 
as I wrote in the new preliminary comment above:
>                     It is now covered by the generalized check_trust()
>                     called above, which checks direct trust for
>                     current cert (at index num_trusted - 1)

DDvO commented on crypto/x509/x509_vfy.c:
```diff
@@ -3054,6 +3047,10 @@ static int build_chain(X509_STORE_CTX *ctx)
             }
             curr = sk_X509_value(ctx->chain, i - 1);
 
+            /* Stop with success if the current cert is trusted */
```
A pity that Stephen Henson did not document the rationale for that behavior.

As stated several times before, the very idea of this PR is to stop chain building as soon as possible, making the resulting chain as short as possible.
This has several advantages:
* IMO more security because only a minimal number of certs are involved in the chain, which less chances of errors/bug/attacks in between.
* efficiency
* sometimes it simply does not make sense to try and find an issuer first, and doing so can cause needless spurious errors

DDvO commented on crypto/x509/x509_vfy.c:
```diff
@@ -3116,31 +3112,39 @@ static int build_chain(X509_STORE_CTX *ctx)
+#if 0 /* This part was used only in two degenerate cases:
+       * * for direct trust of EE cert
+       * * if X509_V_FLAG_TRUSTED_FIRST unset and trust anchor is also in untrusted
+       * It is now covered by the generalized check_trust() called above,
```
I also think that going back to "untrusted first" is not likely to happen,
but if we wanna keep this option we can remove the tentative `#if 0` and `#endif`,
which I tentatively used to verify/demonstrate that it is not necessary (anymore).

mkris86 reviewed (approved):
Geezzeee this is hard

vdukhovni commented on crypto/x509/x509_vfy.c:
```diff
@@ -3094,8 +3091,11 @@ static int build_chain(X509_STORE_CTX *ctx)
-            ok = num > depth ? 0 : get_issuer(&issuer, ctx, curr);
+            /* Stop immediately with success if the current cert is trusted */
+            if ((trust = check_trust(ctx, num)) == X509_TRUST_TRUSTED)
+                break;
```
Didn't we finally agree that this is not quite right? Was that in a separate PR that overlaps with this one?

DDvO replied:
Right. It was in this PR, namely in https://github.com/openssl/openssl/pull/13748#discussion_r581016359 - please place any further comments on that specific topic there.
I just haven't found the time yet to adapt this PR accordingly.

What you could do in the meantime is to answer to my latest reactions to your comments in #13735. This has been waiting for your input since end of January,
and would be great if we can finalize that soon.

DDvO replied:
**Update:** @vdukhovni, I've just done that as discussed above.

vdukhovni commented:
I think we need to set up a voice call and go through all the outstanding open questions in the multiple X509 PRs.  There are just too many places to find all the unresolved questions.  I've lost track of where to look.

My timezone is GMT-0400.  I can do 9AM to 10AM some morning, pick a day and send me a conference link, along with a list of URLs for the relevant PRs and specific unresolved comments inside those PRs.

t8m commented:
Could you please rebase the PR to resolve conflicts?

DDvO commented:
Closing for the time being.
